### PR TITLE
Add typed queue classes

### PR DESCRIPTION
**Pull request by nex3**
These make it easy to, for example, efficiently enqueue incoming
chunked binary data for processing.

### CODE REVIEW

**Comment by nshahan** [comment]
I'm not familiar with this package. 

@eernstg I think you are in the process of migrating dart:typed_data. Would you also want to review this addition to package:typed_data with your API changes in mind?

**Comment by eernstg** [comment]
Hi @nshahan and @nex3,
I think @lrhn would be the obvious choice for such a review&mdash;he has been the maintainer of the core libraries for a long time.

@lrhn, wdyt? We could split the work, too.

**Comment by nex3** [comment]
Ping!

File: lib/src/typed_queue.dart [hunk 1]
```diff
@@ -0,0 +1,608 @@
...
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import "dart:collection";
+import "dart:typed_data";
+
+import "package:collection/collection.dart";
+
+/// The shared superclass of all the typed queue subclasses.
+abstract class _TypedQueue<E, L extends List<E>> extends Object
```
[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/typed_queue.dart]: Done.

File: lib/src/typed_queue.dart [hunk 2]
```diff
@@ -0,0 +1,608 @@
...
+    with ListMixin<E> {
+  /// The underlying data buffer.
+  ///
+  /// This is always both a List<E> and a TypedData, which we don't have a type
+  /// for that. For example, for a `Uint8Buffer`, this is a `Uint8List`.
```
[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/typed_queue.dart]: Done.

File: lib/src/typed_queue.dart [hunk 3]
```diff
@@ -0,0 +1,608 @@
...
+  _TypedQueue(this._table)
+      : _head = 0,
+        _tail = 0;
+
+  // Iterable interface.
+
+  int get length => (_tail - _head) & (_table.length - 1);
+
+  List<E> toList({bool growable = true}) {
+    var list = growable ? (<E>[]..length = length) : _createList(length);
```
[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/typed_queue.dart]: Done.

File: lib/src/typed_queue.dart [hunk 4]
```diff
@@ -0,0 +1,608 @@
...
+    _writeToList(list);
+    return list;
+  }
+
+  QueueList<T> cast<T>() {
+    if (this is QueueList<T>) return this as QueueList<T>;
+    throw UnsupportedError("$this cannot be cast to the desired type.");
+  }
+
+  QueueList<T> retype<T>() => cast<T>();
```
[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/typed_queue.dart]: It exists and is deprecated on `QueueList`:

https://github.com/dart-lang/collection/blob/32f22a541eedd5f7b1bccafade2d316f518f4d5a/lib/src/queue_list.dart#L105-L106

File: lib/src/typed_queue.dart [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+
+  void add(E value) => addLast(value);
+
+  set length(int value) {
+    RangeError.checkNotNegative(value, "length");
+
+    var delta = value - length;
+    if (delta >= 0) {
+      if (_table.length <= value) _growTo(value);
+      _tail = (_tail + delta) & (_table.length - 1);
```
[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/typed_queue.dart]: Done.

File: lib/src/typed_queue.dart [hunk 6]
```diff
@@ -0,0 +1,608 @@
...
+    } else {
+      removeRange(value, length);
+    }
+  }
+
+  E operator [](int index) {
+    RangeError.checkValidIndex(index, this);
```
[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/typed_queue.dart]: Done. That's probably a good note to add to the docs.

File: lib/src/typed_queue.dart [hunk 7]
```diff
@@ -0,0 +1,608 @@
...
+      // [setRange] calls because it could materialize a lazy iterable twice.
+      // Instead we just fall back to the default iteration-based
+      // implementation.
+      super.setRange(start, end, iterable, skipCount);
+    }
+  }
+
+  L sublist(int start, [int end]) {
+    var length = this.length;
+    RangeError.checkValidRange(start, end, length);
```
[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/typed_queue.dart]: Done.

File: lib/src/typed_queue.dart [hunk 8]
```diff
@@ -0,0 +1,608 @@
...
+  ///
+  /// This is functionally identical to `target.setRange(0, end - start, this,
+  /// start)`, but it's more efficient when [target] is typed data.
+  ///
+  /// Returns the number of elements written to [target].
+  int _writeToList(List<E> target, [int start, int end]) {
+    start ??= 0;
+    end ??= length;
+    assert(target.length >= end - start);
+
```
[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/typed_queue.dart]: Done.

File: lib/src/typed_queue.dart [hunk 9]
```diff
@@ -0,0 +1,608 @@
...
+    _tail = _writeToList(newTable);
+    _table = newTable;
+    _head = 0;
+  }
+
+  // Specialization for the specific type.
+
+  // Create a new typed list.
+  L _createList(int size);
+}
```
[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/typed_queue.dart]: I think that could be a good addition to `package:collection` (or `dart:collection`, which implements `QueueList` basically the same way), but unfortunately I don't have the bandwidth for it.

File: lib/src/typed_queue.dart [hunk 10]
```diff
@@ -0,0 +1,608 @@
...
+
+abstract class _IntQueue<L extends List<int>> extends _TypedQueue<int, L> {
```
[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/typed_queue.dart]: We have `_defaultValue` now.

File: test/queue_test.dart [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+      for (var i = capacity ~/ 2 - 1; i > 0; i--) {
+        queue.addFirst(i);
+      }
+      callback(queue);
+    });
+  });
+}
+
+/// Returns a list containing the integers from one through [n].
+List<int> oneThrough(int n) => List.generate(n, (i) => i + 1);
```
[the remark this answers is not in the document]
Comment (nex3) [reply test/queue_test.dart]: I personally prefer `List.generate`.

File: lib/src/typed_queue.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (lrhn) [reply lib/src/typed_queue.dart]: Then also deprecate it here, for great consistency.

[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/typed_queue.dart]: Done.

**Comment by kevmoo** [comment]
What's the story here?

**Comment by nex3** [comment]
Sorry, this dropped off my radar. I'll get it mergeable some time this week.

**Comment by nex3** [comment]
A bunch of new lints were added since I opened this. I've fixed a few one-offs, but the remainders have many failures and no way to do automated fixes. Does someone else care enough to fix them, or can I just comment them out for now?

File: pubspec.yaml [hunk 12]
```diff
@@ -9,6 +9,9 @@ homepage: https://github.com/dart-lang/typed_data
 environment:
   sdk: '>=2.4.0 <3.0.0'
 
+dependencies:
+  collection: ^1.1.0
+
 dev_dependencies:
   pedantic: ^1.0.0
```
Comment (lrhn) [inline pubspec.yaml]: Could also just not depend on pedantic, or lock it to pedantic: 1.7.0 are thereabout.
The newest versions of pedantic have jumped the shark, and I do not want to support the lints they have chose to enable.

File: lib/src/typed_queue.dart [hunk 13]
```diff
@@ -27,17 +27,20 @@ abstract class _TypedQueue<E, L extends List<E>> with ListMixin<E> {
 
   // Iterable interface.
 
+  @override
```
Comment (lrhn) [inline lib/src/typed_queue.dart]: Please don't add @override.
Please don't change quotes.

(So, in general, no to all these changes).

Comment (nex3) [reply lib/src/typed_queue.dart]: Done.

File: pubspec.yaml (hunk 12 above)
Comment (nex3) [reply pubspec.yaml]: Done.

**Pull request merged**